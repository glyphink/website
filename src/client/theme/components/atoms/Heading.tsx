/**
 * @file Atoms - Heading
 * @module website/components/atoms/Heading/component
 */

import type { Props } from '@glyphink/website/theme'
import {
  createElement,
  forwardRef,
  type ForwardedRef,
  type ForwardRefExoticComponent,
  type JSX
} from 'react'
import type HeadingProps from './Heading.props.mts'

/**
 * A section heading.
 *
 * Use `props.size` to change the heading level.
 *
 * - https://developer.mozilla.org/docs/Web/HTML/Element/Heading_Elements
 * - https://developer.mozilla.org/docs/Web/API/HTMLHeadingElement
 *
 * @see {@linkcode ForwardRefExoticComponent}
 * @see {@linkcode HeadingProps}
 *
 * @const {ForwardRefExoticComponent<HeadingProps>} Heading
 */
const Heading: ForwardRefExoticComponent<HeadingProps> = forwardRef<
  HTMLHeadingElement,
  HeadingProps
>(
  /**
   * @param {Props<HeadingProps>} props
   *  Component props
   * @param {ForwardedRef<HTMLHeadingElement>} ref
   *  Forwarded ref
   * @return {JSX.Element}
   *  JSX heading element
   */
  (
    props: Props<HeadingProps>,
    ref: ForwardedRef<HTMLHeadingElement>
  ): JSX.Element => {
    const { size, ...rest } = props
    return createElement('h' + size, { ...rest, ref })
  }
)

Heading.displayName = 'Heading'

export default Heading
