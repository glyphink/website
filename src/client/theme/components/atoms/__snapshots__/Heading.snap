// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:theme/components/atoms/Heading > should render heading element (<h1>) 1`] = `
<h1>
  heading 1
</h1>
`;

exports[`unit:theme/components/atoms/Heading > should render heading element (<h2>) 1`] = `
<h2>
  heading 2
</h2>
`;

exports[`unit:theme/components/atoms/Heading > should render heading element (<h3>) 1`] = `
<h3>
  heading 3
</h3>
`;

exports[`unit:theme/components/atoms/Heading > should render heading element (<h4>) 1`] = `
<h4>
  heading 4
</h4>
`;

exports[`unit:theme/components/atoms/Heading > should render heading element (<h5>) 1`] = `
<h5>
  heading 5
</h5>
`;

exports[`unit:theme/components/atoms/Heading > should render heading element (<h6>) 1`] = `
<h6>
  heading 6
</h6>
`;
